src/cmd/ksh93/edit/history.c

```diff
@@ -959,24 +959,27 @@ static int hist_exceptf(Sfio_t *fp, int type, Sfdisc_t *handle)
 {
     UNUSED(data);
     History_t *hp = (History_t *)handle;
+    int ret;
+
     if (type == SF_WRITE) {
         if (errno == ENOSPC || hp->histwfail++ >= 10) return 0;
         // Write failure could be NFS problem, try to re-open.
         int newfd = open(hp->histname, O_BINARY | O_APPEND | O_CREAT | O_RDWR | O_CLOEXEC,
                          S_IRUSR | S_IWUSR);
         int oldfd = sffileno(fp);
         sh_close(oldfd);
-        if (newfd == -1) {
-            errormsg(SH_DICT, 2, "History file write error-%d %s: file unrecoverable", errno,
-                     hp->histname);
-            return -1;
-        }
+        if (newfd == -1) goto fail;
+
         if (sh_fcntl(newfd, F_DUPFD_CLOEXEC, oldfd) != oldfd) {
             close(newfd);
-            return -1;
+            goto fail;
         }
 
-        fcntl(oldfd, F_SETFD, FD_CLOEXEC);
+        do {
+            ret = fcntl(oldfd, F_SETFD, FD_CLOEXEC);
+            if (ret < 0 && errno != EINTR) goto fail;
+        } while (ret < 0 && errno == EINTR);
```

Comment:
It's not possible for this specific `fcntl()` call to fail with EINTR or to fail at all. Which is why the original author didn't check the return value. It should be sufficient to just do `(void)fcntl(...);`.